content/articles/django-custom-template-tags-and-filters/index.md

```diff
@@ -0,0 +1,258 @@
+
+Let’s get started!
+
+###  Prerequisites
+To follow along it's important that you have:
+1. The ablility to create a simple Django application
+2. Python and Django installed.
+3. A suitable IDE such as Pycharm, VS Code etc.
+
+### Takeaways
+By the end of this tutorial, you will be able to:
+1. Understand the use of custom template tags and filters and how they work.
+2. Create your own customized template tags and filters to suit your needs.
+
+### Getting started.
+Your custom tags and filters must be within a Django app. If you want your tags and filters to be used across your entire project, you may create an app just to accomplish that.
+
+1. Create an app called `Contacts` and add it in your `INSTALLED_APPS`. Inside your app , create a directory called `templatetags` in the same level as the views, models etc.
+2. Make the `templatetags` directory a Python package by creating an `__init__.py` file. Create another file `custom_tags.py` to hold your custom tags and filters.
+
+Your app structure will now look something similar to this:
+
+```python
+.Contact
+├── contact
+│   ├── admin.py
+│   ├── apps.py
+│   
+│   ├── migrations
+│   │   └──  __init__.py
+│   │  
+│   │       
+│   ├── models.py
+│   
+│   ├── templatetags
+│   │  └──  custom_tags.py
+│   │
+│   ├── tests.py
+│   ├── url.py
+│   └── views.py
+├── Contact
+│   ├── asgi.py
+│   ├── settings.py
+│   ├── urls.py
+│   └── wsgi.py
+├── db.sqlite3
+└──  manage.py
+```
+
+You can create as many modules as you can inside the `templatetags` folder. You should, however, note that using the `{% load %}` tag will only work for a particular module name, not the app name.
+
+```python
+from Django import template
+register = template.Library()
+```
+
+To load the tags and filters, we will do the following in our template:
+
+```jinja
+{% load custom_tags %}
+```
+
+### Writing Custom Filters
+Filters are usually written either as `{{value|filter}}` or `{{value|filter:'arg'}}` . This means that the filter will or will not take a argument when acting on a value.
+
+Consider a simple contacts list application showing a contact and when they were created.
+
+The model.py looks as follows:
+
+```Python
+from Django.db import models
+
+class Contact(models.Model):
+    name = models.CharField(max_length=50)
+    date = models.DateField()
+```
+
+The application looks as follows with just a little of styling...
+
+![Contact App](/engineering-education/Django-custom-template-tags-and-filters/image1.png)
+
+Now let's create two simple filters that capitalize all strings in a given name, the first character of a word and also sets a color.
+
+We use `@register` to register our filter and give it a name in Django. Our function takes a name as it's value.
+
+To use our filter we use the line `{% load custom_tags %}` in our templates.
+
+template_tags/custom_tags.py
+
+```python
+from Django import template
+
+register = template.Library()
```

Comment:
Explain your code. Use comments and docstrings to document the code in your article.

Reply:
Done